File: edge.py
from constants import (
    EDGE,
    MINIMUM_TRANSFER_VALUE,
    BEGIN_TRANSFER_VALUE,
    AUTO_ATTACK,
)


class Edge:
    def __init__(self, to_node, from_node, id):
        self.item_type = EDGE
        self.to_node = to_node
        self.from_node = from_node
        self.id = id
        self.on = False
        self.flowing = False
        self.popped = False
        self.update_nodes(True)
        self.state = "one-way"
        self.type = EDGE
        self.effects = set()

    def __str__(self):
        return str(self.id)

    def update_nodes(self, initial=False):
        self.to_node.new_edge(self, "incoming", initial)
        self.from_node.new_edge(self, "outgoing", initial)

    def click(self, clicker, button):
        if button == 1 and self.controlled_by(clicker):
            self.switch()

    def switch(self, specified=None):
        if specified is None:
            self.on = not self.on
        else:
            self.on = specified
        if not self.on:
            self.popped = True

    def update(self):
        if (
            not self.on
            or self.from_node.value < MINIMUM_TRANSFER_VALUE
            or not self.flow_allowed()
        ):
            self.flowing = False
        elif self.from_node.value > BEGIN_TRANSFER_VALUE:
            self.flowing = True

        if self.flowing:
            self.flow()
            if not self.popped:
                self.pop()

    def flow_allowed(self):
        return self.to_node.accept_delivery(self.owner)

    def pop(self):
        self.popped = True
        if not self.contested or not AUTO_ATTACK:
            self.on = False

    def flow(self):
        amount_transferred = self.from_node.send_amount()
        self.delivery(amount_transferred)
        self.from_node.value -= amount_transferred

    def delivery(self, amount):
        self.to_node.delivery(amount, self.from_node.owner)

    def check_status(self):
        pass

    def controlled_by(self, player):
        return self.from_node.owner == player

    def can_be_controlled_by(self, player):
        return self.controlled_by(player)

    @property
    def owned(self):
        return self.duo_ownership and self.to_node.owner == self.from_node.owner

    @property
    def duo_ownership(self):
        return self.to_node.owner is not None and self.from_node.owner is not None

    @property
    def contested(self):
        return self.duo_ownership and self.to_node.owner != self.from_node.owner

    @property
    def owner(self):
        return self.from_node.owner

    @property
    def color(self):
        if self.on:
            if self.flowing:
                return self.from_node.color
            return self.from_node.color
        return (50, 50, 50)

    def opposite(self, node):
        if node == self.from_node:
            return self.to_node
        return self.from_node

File: sharp/edge.cs
using UnityEngine;
using System.Collections.Generic;

public class Edge
{
    public Node ToNode { get; private set; }
    public Node FromNode { get; private set; }
    public int Id { get; private set; }
    public bool On { get; set; }
    public bool Flowing { get; set; }
    public bool Owned { get; set; }
    public bool Contested { get; set; }
    public bool Popped { get; set; }

    private const int MINIMUM_TRANSFER_VALUE = 1; // Replace with your constant
    private const int BEGIN_TRANSFER_VALUE = 2; // Replace with your constant
    private const int TRANSFER_RATE = 1; // Replace with your constant

    public Edge(Node toNode, Node fromNode, int id)
    {
        ToNode = toNode;
        FromNode = fromNode;
        Id = id;
        On = false;
        Flowing = false;
        Owned = false;
        Contested = false;
        Popped = false;
        UpdateNodes();
    }

    public void UpdateNodes()
    {
        ToNode.Incoming.Add(this);
        FromNode.Outgoing.Add(this);
    }

    public void Click(Player clicker, int button)
    {
        if (button == 1 && OwnedBy(clicker))
        {
            Switch();
        }
    }

    public void Switch(bool? specified = null)
    {
        On = specified ?? !On;
    }

    public void Update()
    {
        if (FromNode.Value < MINIMUM_TRANSFER_VALUE || !On || (ToNode.Full && !Contested))
        {
            Flowing = false;
        }
        else if (FromNode.Value > BEGIN_TRANSFER_VALUE)
        {
            Flowing = true;
        }

        if (Flowing)
        {
            Flow();
            if (!Popped)
            {
                Pop();
            }
        }
    }

    public void Pop()
    {
        Popped = true;
        if (!Contested || !FromNode.Owner.AutoAttack)
        {
            On = false;
        }
    }

    public void Flow()
    {
        int amountTransferred = TRANSFER_RATE * FromNode.Value;
        Delivery(amountTransferred);
        FromNode.Value -= amountTransferred;
    }

    public void Delivery(int amount)
    {
        if (ToNode.Owner != FromNode.Owner)
        {
            ToNode.Value -= amount;
            if (ToNode.Killed())
            {
                Capture();
            }
        }
        else
        {
            if (ToNode.Owner == null)
            {
                Capture();
            }
            ToNode.Value += amount;
        }
    }

    public void Capture()
    {
        ToNode.Capture(FromNode.Owner);
    }

    public void CheckStatus()
    {
        Owned = false;
        Contested = false;
        if (ToNode.Owner == null || FromNode.Owner == null)
        {
            return;
        }
        else if (ToNode.Owner == FromNode.Owner)
        {
            Owned = true;
        }
        else
        {
            Contested = true;
        }
    }

    public bool OwnedBy(Player player)
    {
        return FromNode.Owner == player;
    }

    public Color EdgeColor
    {
        get
        {
            if (On)
            {
                return FromNode.NodeColor;
            }
            return new Color(0.2f, 0.2f, 0.2f); // Equivalent to (50, 50, 50)
        }
    }
}
